trio_websocket/__init__.py

```diff
@@ -4,18 +4,90 @@
 import ssl
 from functools import partial
 
+from async_generator import async_generator, yield_, asynccontextmanager
 import trio
 import trio.abc
 import wsproto.connection as wsconnection
 import wsproto.events as wsevents
 import wsproto.frame_protocol as wsframeproto
+from yarl import URL
 
 
 __version__ = '0.2.0-dev'
 RECEIVE_BYTES = 4096
 logger = logging.getLogger('trio-websocket')
 
 
+@asynccontextmanager
+@async_generator
+async def open_websocket(nursery, host, port, resource, use_ssl):
+    '''
+    Open a WebSocket client connection to a host.
+
+    This function is an async context manager that automatically connects and
+    disconnects. It yields a `WebSocketConnection` instance.
+
+    :param nursery: a trio Nursery to run background tasks in
+    :param str host: the host to connect to
+    :param int port: the port to connect to
+    :param str resource: the resource (i.e. path without leading slash)
+    :param use_ssl: a bool or SSLContext
+    '''
+
+    if use_ssl == True:
+        ssl_context = ssl.create_default_context()
+    elif use_ssl == False:
+        ssl_context = None
+    elif isinstance(use_ssl, ssl.SSLContext):
+        ssl_context = use_ssl
+    else:
+        raise TypeError('`use_ssl` argument must be bool or ssl.SSLContext')
+
+    logger.debug('Connecting to ws%s://%s:%d/%s',
+        '' if ssl_context is None else 's', host, port, resource)
+    if ssl_context is None:
+        stream = await trio.open_tcp_stream(host, port)
+    else:
+        stream = await trio.open_ssl_over_tcp_stream(host, port,
+            ssl_context=ssl_context, https_compatible=True)
+    if port in (80, 443):
+        host_header = host
+    else:
+        host_header = '{}:{}'.format(host, port)
+    wsproto = wsconnection.WSConnection(wsconnection.CLIENT, host=host_header,
+        resource=resource)
+    connection = WebSocketConnection(stream, wsproto, path=resource)
+    nursery.start_soon(connection._reader_task)
+    await connection._open_handshake.wait()
+    async with connection:
+        await yield_(connection)
+
+
+@asynccontextmanager
+@async_generator
+async def open_websocket_url(nursery, url, ssl_context=None):
```

Comment:
I believe you can make this a non-async, non-contextmanager function which simply defers to `open_websocket()`  (literally `return open_websocket(...)`)

Reply:
Ahh yes, you're right!